### PR TITLE
Ladybird: JS console improvements

### PR DESCRIPTION
**Pull request by SebastianZaha**

### CODE REVIEW

File: Ladybird/ConsoleWidget.cpp [hunk 1]
```diff
@@ -184,4 +171,48 @@ void ConsoleWidget::reset()
     m_waiting_for_messages = false;
 }
 
+void ConsoleInputEdit::keyPressEvent(QKeyEvent* event)
+{
+    switch (event->key()) {
+    case Qt::Key_Down: {
+        auto last_idx = m_history.size() - 1;
```
Comment (linusg) [inline Ladybird/ConsoleWidget.cpp:178]: Please spell out "index"

File: Ladybird/ConsoleWidget.cpp [hunk 2]
```diff
@@ -159,9 +159,13 @@ void ConsoleWidget::print_html(StringView line)
...
 )~~~"sv);
 
-    // FIXME: Make it scroll to the bottom, using `window.scrollTo()` in the JS above.
-    //        We used to call `m_output_view->scroll_to_bottom();` here, but that does not work because
-    //        it runs synchronously, meaning it happens before the HTML is output via IPC above.
+    // It should be sufficient to scrollTo a y value of document.documentElement.offsetHeight,
+    // but due to an unknown bug offsetHeight seems to not be properly updated after spamming
+    // a lot of document changes.
+    //
+    // The 1ms setTimeout makes the scrollTo async and allows the DOM to be updated.
```
Comment (linusg) [inline Ladybird/ConsoleWidget.cpp:153]: That seems unnecessary, `setTimeout` with a 0 delay is still queued and runs asynchronously

File: Ladybird/ConsoleWidget.cpp [hunk 3]
```diff
@@ -159,9 +159,13 @@ void ConsoleWidget::print_html(StringView line)
         document.body.appendChild(p);
 )~~~"sv);
 
-    // FIXME: Make it scroll to the bottom, using `window.scrollTo()` in the JS above.
-    //        We used to call `m_output_view->scroll_to_bottom();` here, but that does not work because
-    //        it runs synchronously, meaning it happens before the HTML is output via IPC above.
+    // It should be sufficient to scrollTo a y value of document.documentElement.offsetHeight,
```
Comment (linusg) [inline Ladybird/ConsoleWidget.cpp:149]: `// FIXME: `

File: Ladybird/Tab.cpp [hunk 4]
```diff
@@ -667,6 +667,17 @@ void Tab::show_console_window()
         m_console_widget = new Ladybird::ConsoleWidget;
         m_console_widget->setWindowTitle("JS Console");
         m_console_widget->resize(640, 480);
+
+        // Make the copy action available in the window via the bound copy key shortcut.
+        // Simply adding it to the context menu is not enough.
+        m_console_widget->addAction(&m_window->copy_selection_action());
+
+        m_console_context_menu = make<QMenu>("Context menu", m_console_widget);
```
Comment (ADKaster) [inline Ladybird/Tab.cpp:675]: Tab is a QObject, yes? If so, we should use the Qt ownership model here, not OwnPtr.

File: Ladybird/ConsoleWidget.cpp [hunk 5]
```diff
@@ -159,9 +146,13 @@ void ConsoleWidget::print_html(StringView line)
...
 )~~~"sv);
 
-    // FIXME: Make it scroll to the bottom, using `window.scrollTo()` in the JS above.
-    //        We used to call `m_output_view->scroll_to_bottom();` here, but that does not work because
-    //        it runs synchronously, meaning it happens before the HTML is output via IPC above.
+    // FIXME: It should be sufficient to scrollTo a y value of document.documentElement.offsetHeight,
+    // but due to an unknown bug offsetHeight seems to not be properly updated after spamming
+    // a lot of document changes.
+    //
+    // The 1ms setTimeout makes the scrollTo async and allows the DOM to be updated.
```
Comment (linusg) [inline Ladybird/ConsoleWidget.cpp:153]:
```suggestion
```

**Review by linusg — CHANGES_REQUESTED** [review]

File: Ladybird/ConsoleWidget.cpp (hunk 5 above)
Comment (SebastianZaha) [reply Ladybird/ConsoleWidget.cpp]: whoops, thanks.

**Review by linusg — APPROVED** [review]
LGTM, let's wait for a ✅ from @ADKaster

**Comment by SebastianZaha** [comment]
I think you can use a signal only for the Enter keypress, `returnPressed` signal.

That's what I understood at least from [the qt docs](https://doc.qt.io/qt-6/qlineedit.html). But I'm a qt newbie so it's possible I'm missing something :^)

**Comment by SebastianZaha** [comment]
@ADKaster I checked with Valgrind a bit and found some reported leaks with QAction instantiation, added a commit to fix. The OwnPtr members do not seem to be leaking.

There are a bunch of other leaks reported but seem to not be related to ladybird code, or I'm not quite sure how to track them down. (A bunch of them refer to gtk and pango libraries too which is weird, and I suppose related to my window manager?)

**Comment by ADKaster** [comment]
Sounds good. Yeah I wouldn't worry about anything from GTK, Wayland, X11, etc. Without passing the correct suppression files, those libraries will report leaks on their globals

**Pull request merged**